File: trie/find_duplicate_rows_binary_matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

struct TrieNode {
  bool leaf;
  TrieNode* children[2];
  TrieNode() : leaf(false) { children[0] = children[1] = 0; }
};

class Solution {
  void put(TrieNode* p, const vector<int>& line) {
    for (int e : line) {
      if (!p->children[e]) {
        p->children[e] = new TrieNode();
      }
      p = p->children[e];
    }
    p->leaf = true;
  }

  bool get(TrieNode* p, const vector<int>& line) {
    for (int e : line) {
      if (!p->children[e]) return false;
      p = p->children[e];
    }
    return p->leaf;
  }

 public:
  vector<int> repeatedRows(vector<vector<int>> matrix, int M, int N) {
    // Your code here
    TrieNode* root = new TrieNode();
    vector<int> ans;
    for (int i = 0; i < M; ++i) {
      if (get(root, matrix[i])) {
        ans.push_back(i);
      } else {
        put(root, matrix[i]);
      }
    }
    return ans;
  }
};

//{ Driver Code Starts.
int main() {
  int t;
  cin >> t;
  while (t--) {
    int row, col;
    cin >> row >> col;
    vector<vector<int>> matrix(row);
    for (int i = 0; i < row; i++) {
      matrix[i].assign(col, 0);
      for (int j = 0; j < col; j++) {
        cin >> matrix[i][j];
      }
    }

    Solution ob;
    vector<int> ans = ob.repeatedRows(matrix, row, col);

    for (int i = 0; i < ans.size(); ++i) {
      cout << ans[i] << " ";
    }
    cout << endl;
  }
  return 0;
}

// } Driver Code Ends

File: trie/find_duplicate_rows_binary_matrix.py
# User function Template for python3


class TrieNode:
    def __init__(self):
        self.children = [None, None]


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def find(self, a):
        n = len(a)
        p = self.root
        i = 0
        while i < n:
            if p.children[a[i]] is None:
                return False
            p = p.children[a[i]]
            i += 1
        return True

    def insert(self, a):
        n = len(a)
        p = self.root
        i = 0
        while i < n:
            if p.children[a[i]] is None:
                p.children[a[i]] = TrieNode()
            p = p.children[a[i]]
            i += 1


class Solution:
    def repeatedRows(self, arr, m, n):
        # code here
        ans = []
        root = Trie()
        for i in range(m):
            if root.find(arr[i]):
                ans.append(i)
            root.insert(arr[i])

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    tc = int(input())
    while tc > 0:
        n, m = list(map(int, input().strip().split()))

        arr = []
        for i in range(n):
            nums = list(map(int, input().strip().split()))
            arr.append(nums)
        ob = Solution()
        ans = ob.repeatedRows(arr, n, m)

        for x in ans:
            print(x, end=" ")

        print()
        tc -= 1

# } Driver Code Ends
